feat: cache attribute symbols

TimothyMakkison opened the pull request:
**Description**
See #475

Add a `Dictionary<Type, INamedTypeSymbol?>` cache to `WellKnownTypes` to optimize `AttributeDataAccessor`.

**Benchmarks**
Benefits mappers with many mapping methods, saves 8ms for `LargeCompile` and saved 300 KB of memory. Doesn't look like `Compile` is effected.

**Old:**

|       Method |      Mean |     Error |    StdDev |      Gen0 |     Gen1 |  Allocated |
|------------- |----------:|----------:|----------:|----------:|---------:|-----------:|
|      Compile |  1.563 ms | 0.0261 ms | 0.0244 ms |   82.0313 |  19.5313 |  169.32 KB |
| LargeCompile | 59.450 ms | 1.1351 ms | 1.1148 ms | 1428.5714 | 285.7143 | 9280.05 KB |

|       Method |      Mean |     Error |    StdDev |      Gen0 |     Gen1 |  Allocated |
|------------- |----------:|----------:|----------:|----------:|---------:|-----------:|
|      Compile |  1.536 ms | 0.0238 ms | 0.0211 ms |   82.0313 |  19.5313 |  170.23 KB |
| LargeCompile | 57.525 ms | 0.5249 ms | 0.5155 ms | 1428.5714 | 285.7143 | 9276.68 KB |

|       Method |      Mean |     Error |    StdDev |      Gen0 |     Gen1 |  Allocated |
|------------- |----------:|----------:|----------:|----------:|---------:|-----------:|
|      Compile |  1.571 ms | 0.0240 ms | 0.0225 ms |   78.1250 |  15.6250 |  169.85 KB |
| LargeCompile | 58.310 ms | 0.9858 ms | 1.2819 ms | 1428.5714 | 285.7143 | 9272.75 KB |


**New:**


|       Method |      Mean |     Error |    StdDev |      Gen0 |     Gen1 |  Allocated |
|------------- |----------:|----------:|----------:|----------:|---------:|-----------:|
|      Compile |  1.594 ms | 0.0303 ms | 0.0384 ms |   78.1250 |  15.6250 |  168.44 KB |
| LargeCompile | 51.705 ms | 0.8099 ms | 1.1616 ms | 1375.0000 | 375.0000 | 8934.55 KB |

|       Method |      Mean |     Error |    StdDev |      Gen0 |     Gen1 |  Allocated |
|------------- |----------:|----------:|----------:|----------:|---------:|-----------:|
|      Compile |  1.575 ms | 0.0284 ms | 0.0304 ms |   89.8438 |  23.4375 |  168.49 KB |
| LargeCompile | 51.513 ms | 0.7022 ms | 1.1924 ms | 1333.3333 | 333.3333 | 8965.35 KB |

|       Method |      Mean |     Error |    StdDev |      Gen0 |     Gen1 | Allocated |
|------------- |----------:|----------:|----------:|----------:|---------:|----------:|
|      Compile |  1.559 ms | 0.0296 ms | 0.0277 ms |   78.1250 |  19.5313 | 169.52 KB |
| LargeCompile | 51.809 ms | 0.9939 ms | 1.0207 ms | 1375.0000 | 375.0000 | 8925.2 KB |

latonz commented on src/Riok.Mapperly/Descriptors/WellKnownTypes.cs:
```diff
@@ -9,6 +9,7 @@ namespace Riok.Mapperly.Descriptors;
     private readonly Compilation _compilation;
+    private readonly Dictionary<string, INamedTypeSymbol?> _cachedTypes = new();
 
     private INamedTypeSymbol? _referenceHandlerAttribute;
```
I think I'd remove all those fields and always use `GetTypeSymbol` as it should be constant time and pretty fast now and the fields grew a lot over the past.

TimothyMakkison replied:
Okay I'll replace the fields with `GetTypeSymbol(typeof(ImmutableList<>))` etc.
Would it be worth keeping the heavily used ones like `ICollectionT`?

latonz reviewed (changes requested):
Thank you for this improvement 😊

TimothyMakkison commented:
Removed all fields from `WellKnownTypes`, removed nullables and added `TryGetTypeSymbol` for `DateOnly` and `TimeOnly`. `WellKnownTypes` is now instantiated in `MapperGenerator` and shared between instances saving another 400KB for LargeCompile. 

The DI passes although I'm a little concerned I've broken .NET Framework by removing some of the null checks.

Ran the Benchmarks a couple of times, a little more variance but I think thats a combination of the weather and using the most recent branch of Mapperly.

**Old:**

|       Method |      Mean |     Error |    StdDev |      Gen0 |     Gen1 | Allocated |
|------------- |----------:|----------:|----------:|----------:|---------:|----------:|
|      Compile |  1.544 ms | 0.0306 ms | 0.0301 ms |   78.1250 |  15.6250 | 170.25 KB |
| LargeCompile | 59.269 ms | 1.0096 ms | 1.7144 ms | 1500.0000 | 500.0000 | 9276.7 KB |

|       Method |      Mean |     Error |    StdDev |    Median |      Gen0 |     Gen1 |  Allocated |
|------------- |----------:|----------:|----------:|----------:|----------:|---------:|-----------:|
|      Compile |  1.564 ms | 0.0301 ms | 0.0358 ms |  1.554 ms |   82.0313 |  15.6250 |  168.29 KB |
| LargeCompile | 60.973 ms | 1.2988 ms | 3.5774 ms | 59.640 ms | 1428.5714 | 428.5714 | 9274.68 KB |

|       Method |      Mean |     Error |    StdDev |      Gen0 |     Gen1 |  Allocated |
|------------- |----------:|----------:|----------:|----------:|---------:|-----------:|
|      Compile |  1.553 ms | 0.0285 ms | 0.0267 ms |   89.8438 |  19.5313 |  172.33 KB |
| LargeCompile | 58.498 ms | 0.9764 ms | 1.1991 ms | 1428.5714 | 285.7143 | 9266.14 KB |

**New:**
|       Method |      Mean |     Error |    StdDev |    Median |      Gen0 |     Gen1 |  Allocated |
|------------- |----------:|----------:|----------:|----------:|----------:|---------:|-----------:|
|      Compile |  1.852 ms | 0.0990 ms | 0.2856 ms |  1.721 ms |   78.1250 |  19.5313 |  171.06 KB |
| LargeCompile | 48.584 ms | 0.9688 ms | 0.8589 ms | 48.693 ms | 1333.3333 | 333.3333 | 8531.57 KB |

|       Method |      Mean |     Error |     StdDev |    Median |      Gen0 |     Gen1 |  Allocated |
|------------- |----------:|----------:|-----------:|----------:|----------:|---------:|-----------:|
|      Compile |  2.059 ms | 0.1961 ms |  0.5783 ms |  1.740 ms |   70.3125 |  15.6250 |  168.78 KB |
| LargeCompile | 63.516 ms | 6.1086 ms | 18.0113 ms | 57.057 ms | 1285.7143 | 285.7143 | 8532.13 KB |

|       Method |      Mean |     Error |    StdDev |    Median |      Gen0 |     Gen1 | Allocated |
|------------- |----------:|----------:|----------:|----------:|----------:|---------:|----------:|
|      Compile |  2.168 ms | 0.1911 ms | 0.5634 ms |  1.989 ms |   85.9375 |  15.6250 | 170.77 KB |
| LargeCompile | 49.377 ms | 0.9812 ms | 2.0482 ms | 49.311 ms | 1333.3333 | 500.0000 | 8534.1 KB |

latonz commented on src/Riok.Mapperly/Descriptors/MappingBodyBuilders/NewInstanceObjectMemberMappingBodyBuilder.cs:
```diff
@@ -181,15 +181,15 @@ private static void BuildConstructorMapping(INewInstanceBuilderContext<IMapping>
-            .OrderByDescending(x => x.HasAttribute(ctx.BuilderContext.Types.MapperConstructorAttribute))
-            .ThenBy(x => x.HasAttribute(ctx.BuilderContext.Types.ObsoleteAttribute))
+            .OrderByDescending(x => x.HasAttribute(ctx.BuilderContext.Types.GetTypeSymbol(typeof(MapperConstructorAttribute))))
+            .ThenBy(x => x.HasAttribute(ctx.BuilderContext.Types.GetTypeSymbol(typeof(ObsoleteAttribute))))
             .ThenByDescending(x => x.Parameters.Length == 0)
             .ThenByDescending(x => x.Parameters.Length);
         foreach (var ctorCandidate in ctorCandidates)
         {
             if (!TryBuildConstructorMapping(ctx, ctorCandidate, out var mappedTargetMemberNames, out var constructorParameterMappings))
             {
-                if (ctorCandidate.HasAttribute(ctx.BuilderContext.Types.MapperConstructorAttribute))
+                if (ctorCandidate.HasAttribute(ctx.BuilderContext.Types.GetTypeSymbol(typeof(MapperConstructorAttribute))))
```
Couldn't the generic overloads be used here?

latonz commented on src/Riok.Mapperly/Descriptors/WellKnownTypes.cs:
```diff
@@ -1,127 +1,54 @@
-    private INamedTypeSymbol? _dictionaryT;
-    private INamedTypeSymbol? _enum;
-    private INamedTypeSymbol? _type;
+    public INamedTypeSymbol GetTypeSymbol(Type type) =>
```
I think these method names could be simplified to `Get`/`TryGet` as I think it should be clear that it is about types as the class is named `WellKnownTypes`.

TimothyMakkison commented:
Thanks for the review! I've resued the logic between `Get` and `TryGet` and used the generic overload where possible.
